me0Muons in 75X

dnash86 opened the pull request:
Functionality for the reco::ME0Muon objects ported from 620_SLHCX.

Does the "Can't automatically merge" indicate that there's a conflict that must be resolved?  (Maybe in the BuildFiles?)

dnash86 committed:
First commit, compilation working

dnash86 committed:
Upload of analyzer, commented

dnash86 committed:
Updates needed for assoc by hits in analyzer

cmsbuild commented:
A new Pull Request was created by @dnash86  for CMSSW_7_5_X.

me0Muons in 75X

It involves the following packages:

DataFormats/GEMRecHit
DataFormats/MuonReco
RecoMuon/MuonIdentification
Validation/RecoMuon


@cmsbuild, @cvuosalo, @danduggan, @deguio, @slava77 can you please review it and eventually sign? Thanks.
@battibass, @abbiendi, @jhgoh, @bellan, @trocino, @amagitte, @bachtis, @rociovilar this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.cc:
```diff
@@ -0,0 +1,121 @@
+
+      const reco::Track* trk = &(*trackref);
+      // pointer to old track:
+      reco::Track* newTrk = new reco::Track(*trk);
```
what's the point of making a copy?
This is a memory leak

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.h:
```diff
@@ -0,0 +1,35 @@
+#include "DataFormats/DTRecHit/interface/DTRecSegment4DCollection.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+
+class ME0MuonTrackCollProducer : public edm::EDProducer {
```
all new modules should be using thread - aware base classes.
edm::stream::EDProducer appears to be appropriate

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.cc:
```diff
@@ -0,0 +1,121 @@
+
+#include <sstream>
+
+// std::vector<double> ME0MuonTrackCollProducer::findSimVtx(edm::Event& iEvent){
```
please remove commented out code or explain why it's needed here

slava77 commented:
RecoMuon/MuonIdentification/test/ME0MuonAnalyzer.cc is currently not showing up in diffs on the web because it's so large. To aid review, please reduce the size of the file (there appears to be enough commented out code there that the reduction would be trivial)

slava77 commented:
The PR currently has merge conflicts due to muon upgrade code integration that happened after 750pre5 (the start of your developments).
Please use the latest IB to update: rebase your topic branch onto it first , if that's not working out for you, you may want to submit a new PR

dnash86 committed:
Updates requested

dnash86 commented:
Updated with fixes to files as requested.  Removed old commented code in RecoMuon/MuonIdentification/test/ME0MuonAnalyzer.cc, left just one small section of commented code for potential future use.  Will work on fixing merge conflicts now.

cmsbuild commented:
Pull request #9556 was updated. @cmsbuild, @cvuosalo, @danduggan, @deguio, @slava77 can you please check and sign again.

cmsbuild commented:
Pull request #9556 was updated. @cmsbuild, @cvuosalo, @danduggan, @deguio, @slava77 can you please check and sign again.

dnash86 commented:
Tried to fix merge conflicts, it seems possible to merge now.  Let me know if there are further problems/questions.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: aec4f1dcd6df95a2edb606984ebbd3cf1314eceb
 I found an error when building: 
 
 <pre>ME0Muon.cc:(.text._ZNK4reco7ME0Muon10me0segmentEv[_ZNK4reco7ME0Muon10me0segmentEv]+0x21): undefined reference to `vtable for ME0Segment'
ME0Muon.cc:(.text._ZNK4reco7ME0Muon10me0segmentEv[_ZNK4reco7ME0Muon10me0segmentEv]+0xb1): undefined reference to `vtable for ME0RecHit'
tmp/slc6_amd64_gcc491/src/DataFormats/MuonReco/src/DataFormatsMuonReco/ME0Muon.o: In function `std::vector<ME0RecHit, std::allocator<ME0RecHit> >::operator=(std::vector<ME0RecHit, std::allocator<ME0RecHit> > const&)':
ME0Muon.cc:(.text._ZNSt6vectorI9ME0RecHitSaIS0_EEaSERKS2_[_ZNSt6vectorI9ME0RecHitSaIS0_EEaSERKS2_]+0x1b1): undefined reference to `vtable for ME0RecHit'
ME0Muon.cc:(.text._ZNSt6vectorI9ME0RecHitSaIS0_EEaSERKS2_[_ZNSt6vectorI9ME0RecHitSaIS0_EEaSERKS2_]+0x259): undefined reference to `vtable for ME0RecHit'
collect2: error: ld returned 1 exit status
gmake: *** [tmp/slc6_amd64_gcc491/src/DataFormats/MuonReco/src/DataFormatsMuonReco/libDataFormatsMuonReco.so] Error 1
Leaving library rule at Calibration/EcalAlCaRecoProducers
Entering library rule at RecoPixelVertexing/PixelTrackFitting
>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_7_5_X_2015-06-11-1100/src/RecoPixelVertexing/PixelTriplets/src/CosmicHitTripletGenerator.cc 
Leaving library rule at DataFormats/MuonReco
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/5446/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/aec4f1dcd6df95a2edb606984ebbd3cf1314eceb
https://github.com/cms-sw/cmssw/commit/1f9d2fb3cc04ae0994a2b8126bdf1de49fea6fad
https://github.com/cms-sw/cmssw/commit/906af7080b2b3ddc90853a45b4e364bf362fb491
https://github.com/cms-sw/cmssw/commit/6bdd448b720462df6fc08125a2366dac0d5fe43a
https://github.com/cms-sw/cmssw/commit/7fa16dbd765728724c6631979d9fc35533e63f49
https://github.com/cms-sw/cmssw/commit/7cb43c4db1902a4f9b72048ac0ea62b73875da00
https://github.com/cms-sw/cmssw/commit/2401b92109c3487f8454adf001314040ea9b4e08
https://github.com/cms-sw/cmssw/commit/1510b656cae91d552536e470583a892741ff8a8c
https://github.com/cms-sw/cmssw/commit/648a9c98f3a154eac45074cf2e33b21a721ff36a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/5446/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/5446/git-merge-result

deguio commented:
hello @dnash86 
is the ME0MuonTrackCollProducer used? I don't see it included in any sequence and I am wondering why it leaves under Validation
thanks,
F.

dnash86 committed:
Missing classes files

dnash86 committed:
Commit other missing files

dnash86 committed:
Merged CMSSW_7_5_X from repository cms-sw

dnash86 committed:
Fix conflicts.

dnash86 commented:
Committed missing files causing the build failure.  

@deguio - The ME0MuonTrackCollProducer is used in some test analyzer code.  It was helpful for us to have in the release.  Is it okay to keep?  We put it in Validation/RecoMuon since other muon related analyzers are there.

cmsbuild commented:
Pull request #9556 was updated. @cmsbuild, @cvuosalo, @danduggan, @deguio, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 23f1659babb51186e3af8efe1db79e4e386b1f0c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/8172/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/23f1659babb51186e3af8efe1db79e4e386b1f0c
https://github.com/cms-sw/cmssw/commit/1db98159247e426a4b771f70b6c0a21d30277575
https://github.com/cms-sw/cmssw/commit/c439569484699562dad1df8207eb383942925f09
https://github.com/cms-sw/cmssw/commit/9f68ee03a122e434a3b4e70594ed8fd5ccea9d81
https://github.com/cms-sw/cmssw/commit/eb6ebd427fccec4af9ca237e970210c30a03f6cd
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/8172/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/8172/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/8172/summary.html

slava77 commented:
David, please address issues raised by the static code analysis
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9556/8172/llvm-analysis/
- ME0MuonConverter should be thread-aware module ( ``` edm::stream::EDProducer<> ``` seems appropriate)
- ME0MuonSelector non-const static  selectionTypeStringToEnumMap should be made const or recoded non-static
- EmulatedME0Segment, ME0Segment non-const static  theProjectionMatrix and isInitialized  should be made const or recoded non-static

deguio commented:
I don't particularly like having unused classes in release, but I let @davidlange6 comment further.
ciao,
F.

slava77 commented:
@deguio @dnash86 
I think that code from ME0MuonTrackCollProducer should be in the Reco or CommonTools packages.
Its purpose is to pick muons with some quality and write a matching track collection.
There is no validation or monitoring here.

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+ *  \author Marcello Maggi
+ */
+
+#include <DataFormats/TrackingRecHit/interface/RecSegment.h>
```
Please use ``` "DataFormats/TrackingRecHit/interface/RecSegment.h" ``` for regular includes.
The brakets are for system includes.

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+ *  This is 4-dimensional since it has an origin (x,y) and a direction (x,y)
+ *  in the local coordinate system of the chamber.
+ *
+ *  $Date: 2014/02/04 12:41:32 $
```
``` $ some text $ ``` lines are remnants of CVS formatting and should be removed.

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+ ME0Segment() : theChi2(0.){}
+  
+  /// Constructor
+  ME0Segment(const std::vector<const ME0RecHit*>& proto_segment, LocalPoint origin, 
```
pass complex arguments by const reference
``` const LocalPoint&, const LocalVector&, const AlgebraicSymMatrix&, double ```

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+  //--- Base class interface
+  ME0Segment* clone() const { return new ME0Segment(*this); }
+
+  LocalPoint localPosition() const { return theOrigin; }
```
``` const LocalPoint& localPosition() const ```

slava77 replied:
uhm, the return by value is from the base class interface (TrackingRecHit.h).
So, correction:
change to ``` virtual localPosition() const ```
same for all other inherited virtual methods

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+  AlgebraicVector parameters() const;
+
+  /// Covariance matrix of parameters()
+  AlgebraicSymMatrix parametersError() const { return theCovMatrix; }
```
``` const  AlgebraicSymMatrix parametersError() const ```

slava77 replied:
as above ``` virtual AlgebraicSymMatrix parametersError() ```

slava77 commented on DataFormats/GEMRecHit/interface/ME0Segment.h:
```diff
@@ -0,0 +1,85 @@
+
+  virtual std::vector<TrackingRecHit*> recHits();
+
+  double chi2() const { return theChi2; };
```
virtual

slava77 commented on DataFormats/GEMRecHit/interface/ME0SegmentCollection.h:
```diff
@@ -0,0 +1,24 @@
+ *
+ * The collection of ME0Segment's. See \ref CSCSegmentCollection.h for details from which is derived.
+ *
+ *  $Date: 2014/02/04 10:08:15 $
```
remove ```$Date: 2014/02/04 10:08:15 $```

slava77 commented on DataFormats/GEMRecHit/src/ME0Segment.cc:
```diff
@@ -0,0 +1,111 @@
+/** \file ME0egment.cc
+ *
+ *  $Date: 2014/02/04 12:41:33 $
```
remove ```$Date: 2014/02/04 12:41:33 $ ```

slava77 commented on DataFormats/GEMRecHit/src/ME0Segment.cc:
```diff
@@ -0,0 +1,111 @@
+
+
+AlgebraicMatrix ME0Segment::projectionMatrix() const {
+  static AlgebraicMatrix theProjectionMatrix( 4, 5, 0);
```
this should be implemented using const static
see e.g. https://github.com/cms-sw/cmssw/blob/CMSSW_7_5_X/DataFormats/CSCRecHit/src/CSCSegment.cc#L82

slava77 commented on DataFormats/GEMRecHit/src/ME0Segment.cc:
```diff
@@ -0,0 +1,111 @@
+    const  ME0RecHit *recHit = &(*itRH);
+    averageTime+=recHit->tof();
+  }
+  averageTime=averageTime/(theME0RecHits.size());
```
check that the size is not zero

slava77 commented on DataFormats/MuonReco/BuildFile.xml:
```diff
@@ -5,8 +5,9 @@
 <use   name="DataFormats/CSCRecHit"/>
 <use   name="DataFormats/VertexReco"/>
-
+<use   name="rootrflx"/>
```
remove rootrflx dependency here

slava77 commented on DataFormats/MuonReco/BuildFile.xml:
```diff
@@ -5,8 +5,9 @@
-
+<use   name="rootrflx"/>
 <use   name="rootmath"/>
+<use   name="DataFormats/GEMRecHit"/>
```
just for uniformity, please move this up closer to CSCRecHit

slava77 commented on DataFormats/MuonReco/interface/EmulatedME0Segment.h:
```diff
@@ -0,0 +1,77 @@
+ *  It is modeled after CSCSegment, so it is a 4-dimensional object ( origin (x,y) , direction (x,y) )
+ *  Formally it must be defined as a LocalPoint and LocalError but we actually use the global coordinate system.
+ *  
+ *  $Date: 2013/04/22 22:41:32 $
```
remove ``` $Date: 2013/04/22 22:41:32 $ ```

slava77 commented on DataFormats/MuonReco/interface/EmulatedME0Segment.h:
```diff
@@ -0,0 +1,77 @@
+    //--- Base class interface
+    EmulatedME0Segment* clone() const { return new EmulatedME0Segment(*this); }
+
+    LocalPoint localPosition() const { return theOrigin; }
```
all comments about addition of virtual made in ME0Segment.h
 apply here as well

slava77 commented on DataFormats/MuonReco/interface/EmulatedME0SegmentCollection.h:
```diff
@@ -0,0 +1,24 @@
+ *
+ * The collection of EmulatedME0Segment's. See \ref EmulatedME0SegmentCollection.h for details.
+ *
+ *  $Date: 2010/03/12 13:08:15 $
```
drop ``` $ ... $ ```

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+    virtual ~ME0Muon(){}     
+    
+    /// reference to Track reconstructed in the tracker only
+    virtual TrackRef innerTrack() const { return innerTrack_; }
```
why are all of these methods virtual?

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+    /// set reference to our new ME0Segment type
+    virtual void setME0Segment( const ME0Segment & s ) { me0Segment_ = s; }
+
+    virtual ME0Segment me0segment() const { return me0Segment_; }
```
return by const ref ```const ME0Segment& ```

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+    /* void setYdiff( const double ydiff ) { ydiff_ = ydiff; } */
+    /* void setPhidirdiff( const double phidirdiff ) { phidirdiff_ = phidirdiff; } */
+
+    GlobalPoint globalTrackPosAtSurface() const { return globalTrackPosAtSurface_; }
```
use ```const &``` for return values

deguio commented:
thanks @slava77 for your comment.
I would conclude that the DQM/Validation packages are not touched by this PR.

ciao,
F.

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+    /// pseudorapidity of momentum vector
+    double eta() const { return innerTrack_.get()->eta(); }
+
+    //functions for easy variation of me0muon criteria
```
please remove commented out code

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+
+namespace reco {
+ 
+  class ME0Muon {
```
Is there a good reason for this class not to inherit from a RecoCandidate (same as reco::Muon)?

slava77 commented on DataFormats/MuonReco/src/EmulatedME0Segment.cc:
```diff
@@ -0,0 +1,67 @@
+
+
+AlgebraicMatrix EmulatedME0Segment::projectionMatrix() const {
+  static AlgebraicMatrix theProjectionMatrix( 4, 5, 0);
```
this should be recoded to not use non-const statics

slava77 commented on DataFormats/MuonReco/interface/ME0Muon.h:
```diff
@@ -0,0 +1,102 @@
+  public:
+    ME0Muon();
+    //ME0Muon( const TrackRef & t, const ME0Segment & s) { innerTrack_ = t; me0Segment_ = s;}
+    ME0Muon( const TrackRef & t, const ME0Segment & s, const int v) { innerTrack_ = t; me0Segment_ = s; me0segid_=v;}
```
all basic type data members should be initialized (trackCharge is missing)

slava77 commented on DataFormats/MuonReco/src/ME0Muon.cc:
```diff
@@ -0,0 +1,10 @@
+#include "DataFormats/MuonReco/interface/ME0Muon.h"
+using namespace reco;
+
+ME0Muon::ME0Muon() {
```
all basic type data members should be initialized

slava77 commented on DataFormats/MuonReco/interface/EmulatedME0Segment.h:
```diff
@@ -0,0 +1,77 @@
+#ifndef ME0RecHit_EmulatedME0Segment_h
```
I think this should be defined in the same place as the ME0Segment.h (GemRecHit/)

slava77 commented on DataFormats/MuonReco/src/classes_def.xml:
```diff
@@ -177,5 +177,24 @@
+    
+    <class name="std::vector<reco::ME0Muon>"/>
+    <class name="reco::ME0Muon" ClassVersion="17">
+      <version ClassVersion="17" checksum="3425786499"/>
```
I see that 62XSLHC has ME0Muon class version 15
and EmulatedME0Segment version 11.
Was this update in the class version required?

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonConverter.h:
```diff
@@ -0,0 +1,59 @@
+
+
+
+class ME0MuonConverter : public edm::EDProducer {
```
make this ```edm::stream::EDProducer<>```

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonConverter.h:
```diff
@@ -0,0 +1,59 @@
+    /// Destructor
+    ~ME0MuonConverter();
+    /// Produce the converted collection
+    virtual void produce(edm::Event&, const edm::EventSetup&);
```
add ``` override ``` at the end

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonConverter.h:
```diff
@@ -0,0 +1,59 @@
+
+private:
+
+    int iev; // events through
```
is this counter needed?
The producer behavior shouldn't depend on the number of events processed.

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonConverter.h:
```diff
@@ -0,0 +1,59 @@
+/** \class ME0MuonConverter 
+ * Produces a collection of ME0Segment's in endcap muon ME0s. 
+ *
+ * $Date: 2010/03/11 23:48:11 $
```
remove ```$Date: 2010/03/11 23:48:11 $ ```

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonConverter.h:
```diff
@@ -0,0 +1,59 @@
+
+#include "DataFormats/Math/interface/AlgebraicROOTObjects.h"
+#include "TH1.h" 
+#include "TFile.h"
```
These 3 includes appear to be irrelevant. Please remove.

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0MuonSelector.cc:
```diff
@@ -0,0 +1,121 @@
+  bool X_MatchFound = false, Y_MatchFound = false, Dir_MatchFound = false;
+  
+  
+  if ( ( (fabs(thisPosition.x()-r3FinalReco.x())/sigmax ) < MaxPullX ) || (fabs(thisPosition.x()-r3FinalReco.x()) < MaxDiffX ) ) X_MatchFound = true;
```
Why is all of this matching not stored in the ME0Muon itself?
Requirements to have access to geometry essentially block access to these selectors at analysis level.

dnash86 replied:
Possible strategy: Save the local track state and covariance matrix in the ME0Muon object (when they are produced in the ME0SegmentMatcher), and use them in the ME0MuonSelector without needing geometry access.  Does this sound okay to you?

slava77 replied:
You can go to this full grain details, yes.
Just saving of the residuals (deltaX, deltaY, pullX, pullY, and deltaPhiDirection [names are somewhat free-form]) would be enough.

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+#include "DataFormats/GeometrySurface/interface/LocalError.h"
+
+
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
```
this appears to be not needed. Please, remove

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+
+ME0SegmentMatcher::ME0SegmentMatcher(const edm::ParameterSet& pas) : iev(0){
+  produces<std::vector<reco::ME0Muon> >();  
+  X_PULL_CUT   = pas.getParameter<double>("maxPullX");
```
use lower case (bouncing cap, possibly with "the" or an underscore) for data members

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+  PHIDIR_RESIDUAL_CUT   = pas.getParameter<double>("maxDiffPhiDirection");
+  //Might need to replace "OurSegments" with an edm::InputTag of "OurSegments"
+  edm::InputTag OurSegmentsTag ("OurSegments");
+  edm::InputTag generalTracksTag ("generalTracks");
```
is there a good reason for this to be hardcoded?

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+      //Initializing our plane
+
+      //Remove later
+      if (fabs(thisTrack->eta()) < 1.8) continue;
```
std::abs is a better choice than fabs

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+      //Remove later
+      if (fabs(thisTrack->eta()) < 1.8) continue;
+
+      float zSign  = thisTrack->pz()/fabs(thisTrack->pz());
```
``` float zSign = thisTrack->pz() > 0 ? 1.0f : -1.0f;```

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+
+      float zSign  = thisTrack->pz()/fabs(thisTrack->pz());
+
+      float zValue = 526.75 * zSign;
```
change this to a constant, and better start using actual ME0 geometry

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+	
+      //lastrecostate = ThisshProp->propagate(startrecostate, *plane);
+      //lastrecostate = ThisshProp->propagateWithPath(startrecostate, *plane);
+      ThisshProp->propagate(startrecostate, *plane,lastrecostate);
```
I don't see a need to cast to SteppingHelixPropagator and use this non-standard interface.
Use Propagator base class methods instead.

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+	 if ( (fabs(thisPosition.x()-r3FinalReco.x()) < (X_PULL_CUT * sigmax)) || (fabs(thisPosition.x()-r3FinalReco.x()) < X_RESIDUAL_CUT ) ) X_MatchFound = true;
+	 if ( (fabs(thisPosition.y()-r3FinalReco.y()) < (Y_PULL_CUT * sigmay)) || (fabs(thisPosition.y()-r3FinalReco.y()) < Y_RESIDUAL_CUT ) ) Y_MatchFound = true;
+
+	 if ( fabs(p3FinalReco_glob.phi()-roll->toGlobal(thisSegment->localDirection()).phi()) < PHIDIR_RESIDUAL_CUT) Dir_MatchFound = true;
```
did you check that this does 2*pi wraparound correctly?
otherwise, use reco::deltaPhi

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.cc:
```diff
@@ -0,0 +1,285 @@
+	   GlobalPoint SegPos(roll->toGlobal(thisSegment->localPosition()));
+	   GlobalPoint TkPos(r3FinalReco_globv.x(),r3FinalReco_globv.y(),r3FinalReco_globv.z());
+	   
+	   double thisDelR = reco::deltaR(SegPos,TkPos);
```
use ```deltaR2``` instead, there is no need for the sqrt here (deltaR proper is not saved

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.h:
```diff
@@ -0,0 +1,86 @@
+#include "TFile.h"
+#include <TProfile.h>
+#include "TStyle.h"
+#include <TCanvas.h>
```
TH*,  TProfile etc includes are not needed. Please remove.

slava77 commented on RecoMuon/MuonIdentification/plugins/ME0SegmentMatcher.h:
```diff
@@ -0,0 +1,86 @@
+
+private:
+
+    int iev; // events through
```
it looks like this one is not used; please remove

slava77 commented on RecoMuon/MuonIdentification/python/me0MuonConverter_cfi.py:
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+#process = cms.Process("ME0MuonConversion")
+#process.me0MuonConverter = cms.EDProducer("ME0MuonConverter")
```
remove the process* lines

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.cc:
```diff
@@ -0,0 +1,77 @@
+  muonsTag(parset.getParameter< edm::InputTag >("muonsTag")),
+  vxtTag(parset.getParameter< edm::InputTag >("vxtTag")),
+  useIPxy(parset.getUntrackedParameter< bool >("useIPxy", true)),
+  useIPz(parset.getUntrackedParameter< bool >("useIPz", true)),
```
useIP* are apparently not used.
Please remove

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.cc:
```diff
@@ -0,0 +1,77 @@
+  parset_(parset)
+{
+  produces<reco::TrackCollection>();
+  edm::InputTag OurMuonsTag ("me0SegmentMatching");
```
does this have to be hardcoded?

slava77 commented on Validation/RecoMuon/plugins/ME0MuonTrackCollProducer.h:
```diff
@@ -0,0 +1,35 @@
+    ~ME0MuonTrackCollProducer();
+
+  private:
+    virtual void produce(edm::Event&, const edm::EventSetup&);
```
add override: ```virtual void produce(edm::Event&, const edm::EventSetup&) override```

dnash86 committed:
Merged CMSSW_7_5_X from repository cms-sw

dnash86 committed:
Merged refs/pull/9556/head from repository cms-sw

slava77 commented:
@dnash86 
About ME0MuonAnalyzer, the following are mostly suggestions, but please consider them seriously.
- The analyzer is using only histograms (no TTrees). For this kind of implementation it's more meaningful to write the histogram I/O in DQM format and benefit from ability to process in parallel and merge. 
- Currently, you are running both histogram filling and plot making in a single cmsRun application. This is quite inefficient. It makes more sense to save the file and then run a processing script. (this will also let avoid system calls like mkdir from a cmsRun, which isn't really allowed)

In case of staying with TFile : 
- instead of manipulating TFile directly, switch to using TFileService
- after the plot writing (png files) are split into a separate script, the analyzer can be also made multi-thread aware, use ``` edm::one::EDAnalyzer<edm::one::SharedResources> ``` and add TFileService as a shared resource

slava77 commented:
I'm done with the review.
Issues raised for the main part of the code (excluding the ME0MuonAnalyzer) should be addressed before this can be signed off.

dnash86 committed:
Commit technical updates

dnash86 commented:
@slava77 Thanks a lot for your very detailed review.  I've uploaded a number of your technical comments now, although there are still some strategy comments we are working on (so perhaps do not re test yet, until another commit is made).

dnash86 commented:
Regarding the ME0MuonAnalyzer: 

Thank you for the suggestions, is it okay if we perhaps address those in a later commit?  As you mention the main code is more important.

cmsbuild commented:
Pull request #9556 was updated. @cmsbuild, @cvuosalo, @danduggan, @deguio, @slava77 can you please check and sign again.

slava77 commented:
David, are you done with updates? the last commit is pretty small

dnash86 commented:
Regarding the ME0MuonSelector/ME0Muon:

I replied in-line to your comment there.  The ME0Muon object itself is likely a temporary one, for current results, more advanced selection will be done with the full reco code we are working on as well.

dnash86 commented:
@slava77: No, definitely not done with updates yet.  I'm sorry, I wasn't sure if it was okay to commit changes mid-work here, should I have held off until I had everything?

slava77 commented:
now that I know that you are not done, you can keep updating; but it feels more effective to commit and not push until actually done

dnash86 commented:
Yes, very good point, sorry!  I actually have an additional set of changes (addressing more of your technical comments) that I had forgotten to "git add" that I was about to commit.  I will commit and not push changes till the entire work is complete.

dnash86 commented:
Does the strategy for ME0MuonSelector (inline comment) sound okay to you?

slava77 commented:
On 6/17/15 2:12 PM, dnash86 wrote:
> Does the strategy for ME0MuonSelector (inline comment) sound okay to you?

Yes, it's OK, but less info dumped is better

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9556#issuecomment-112917098>.
>

deguio commented:
-1
ciao @dnash86 we agreed that the ME0MuonTrackCollProducer should go in CommonTools instead of Validation.. right?
thanks,
F.

dnash86 commented:
@deguio Sure, yes, I can move ME0MuonTrackCollProducer to CommonTools.  Which subpackage should it be in?  Perhaps RecoUtils?

slava77 commented:
On 6/18/15 3:45 PM, dnash86 wrote:
> @deguio <https://github.com/deguio> Sure, yes, I can move
> ME0MuonTrackCollProducer to CommonTools. Which subpackage should it be
> in? Perhaps RecoUtils?

In the current state, I think it's better in RecoMuon/MuonIdentification
so that the package dependency tree doesn't grow too much.

If you change the content of ME0Muon in a way that  the isGoodMuon
doesn't need to know about the geometry, then CommonTools/RecoAlgos
should be fine

		

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/9556#issuecomment-113284106>.
>

slava77 commented:
ping
@dnash86 how is the progress here?

slava77 commented:
-1

moving to 76X release cycle.
superseded by  #9808
Since that one is based on the same topic branch (dnash86:ME0Muons-In-75X ) David, you can keep updating your original branch.

slava77 commented:
@dnash86 if you still have an old (pre-)release area set up, I suggest you keep it and make commits there.
(just trying to avoid branch rebasing until it's really required)

dnash86 commented:
@slava77 Working on some testing of things (found a few additional needed changes).  Will try to update as soon as I can.  (Delayed by focus on prep for FR of an analysis paper.)

dnash86 commented:
All changes are ready to be pushed to the branch.  

A question about testing:
Is there a runTheMatrix.py option for the PhaseII upgrade scenario in 75X?

slava77 commented:
I don't think the matrix config is present with the full detector.

muon parts can be rerun with one of the configs
that Piet made

RecoLocalMuon/GEMRecHit/test/SingleMuPt100_cfi_GEM-SIM-DIGI_Extended2023_cfg.py

As far as I know, the Phase 2 tracker will not work.
@pietverwilligen I recall that you made a hybrid geometry setup with run 2 tracker and Muon2023.
.. I can't recall if the ME0 part was made there as well.

dnash86 commented:
@slava77 The point regarding the Phase 2 tracker is an important one - this PR is mostly for a matching scheme between ME0Segments and tracks, so full testing may need to wait till after the Phase2 tracker will work.

slava77 commented:
@dnash86 with run2 tracker you will still be able to test a half of the coverage (up to eta 2.5)

dnash86 commented:
@slava77, that is true, I will try that.

slava77 commented:
ping 
David, please clarify the status.
Thanks

dnash86 commented:
Working on running a test on changes (had to start with me0digis, which had a small problem that had to be resolved).

Current status:  Just now trying to work out how to configure cmsDriver.py to run with SLHC me0Customs from sim through to reco, with this config: SingleMuPt100_cfi.

davidlange6 commented:
@dnash86 
closing this - please reopen in 76x when updated.
